api/src/main/java/org/openmrs/api/TagService.java

```diff
@@ -0,0 +1,156 @@
+	 */
+	@Authorized(TagConstants.MANAGE_TAGS)
+	Tag saveTag(Tag tag);
+	
+	/**
+	 * Completely delete the tag from the database.
+	 * 
+	 * @param tag The Tag to remove from the system
+	 */
+	@Authorized(TagConstants.MANAGE_TAGS)
+	void deleteTag(Tag tag);
+	
+	/**
+	 * Removes a tag(if exists) from an OpenmrsObject.
+	 * 
+	 * @param openmrsObject the openmrsObject from which the tags is to be removed
+	 * @param tag the textual tag to be removed
+	 */
+	@Authorized(TagConstants.MANAGE_TAGS)
+	void removeTag(OpenmrsObject openmrsObject, String tag);
+	
+	/**
+	 * Returns a tag by id.
+	 * 
+	 * @param tagId the id of the Tag Object to be retrieved
+	 * @return Tag with given internal identifier
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	Tag getTag(Integer tagId);
+	
+	/**
+	 * Returns a list of Tag objects having the given tag, or having tags containing the substring.
+	 * Partial Search supported
+	 * 
+	 * @param tag the tag (String) which the
+	 * @return List of Tag Objects
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	List<Tag> getTags(String tag);
+	
+	/**
+	 * Returns a list of All Tags.
+	 * 
+	 * @return list Of Tags (String)
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	List<String> getAllTags();
+	
+	/**
+	 * Retrieves an Object with given object uuid and object type from the Openmrs database.
+	 * 
+	 * @param objectUuid the UUID of the object
+	 * @param objectType the java class of the object (eg. Concept.class)
+	 * @return
+	 */
+	<T extends OpenmrsObject> T getObject(Class<T> objectType, String objectUuid);
+	
+	/**
+	 * It adds a tag to the openmrs object, if such a tag does not already exist on the object. A
+	 * warning will be logged if the object does not exist in the database
+	 * 
+	 * @param openmrsObject the object to which the tag has to be added
+	 * @param tag the tag (string) which is to be added
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	Tag addTag(OpenmrsObject openmrsObject, String tag);
+	
+	/**
+	 * It returns a list of tags added to the openmrs object
+	 * 
+	 * @param openmrsObject the object whose tags needs to obtained
+	 * @return a List of Tag Objects having the same object type and uuid as the Openmrs Object
+	 *         parameter
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	List<Tag> getTags(OpenmrsObject openmrsObject);
+	
+	/**
+	 * returns a List of Tag Objects which have an Object Type that belongs to the objectTypes
+	 * parameter and a tag which belongs to the tags (List<String>) parameter. returns a list of Tag
+	 * Objects which have a tag belonging to the List, if objectTypes list is empty. Throws an API
+	 * Exception if tags list is empty.
+	 * 
+	 * @param objectTypes a list of object types (objectTypes should be the Java Class Name)
+	 * @param tags a list of tags
+	 * @return a list of tag objects which have an object_type and a tag from the parameters
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	List<Tag> getTags(List<String> objectTypes, List<String> tags);
```

Comment:
When searching we want to display the tagged objects and not the tag objects, that's why it actually makes sense to include the OpenmrsObjects themselves in the return values